code/lib/preview-api/src/modules/store/csf/prepareStory.ts

```diff
@@ -105,8 +105,19 @@ export function prepareStory<TRenderer extends Renderer>(
     }
 
     const mappedArgs = Object.entries(finalContext.args).reduce((acc, [key, val]) => {
-      const mapping = finalContext.argTypes[key]?.mapping;
-      acc[key] = mapping && val in mapping ? mapping[val] : val;
+      const mappingFn = finalContext.argTypes[key]?.mapping
+        ? (ele: any) =>
+            ele in finalContext.argTypes[key].mapping
+              ? finalContext.argTypes[key].mapping[ele]
+              : ele
+        : null;
+
+      if (mappingFn) {
+        acc[key] = Array.isArray(val) ? val.map(mappingFn) : mappingFn(val);
+      } else {
+        acc[key] = val;
+      }
+
       return acc;
```

Comment:
I've tried to simplify the logic a bit here, would this still work?

```suggestion
      
      if(!finalContext.argTypes[key]?.mapping)){
        acc[key] = val;
        return acc;
      }
      
      const mappingFn = (originalValue: any) =>
        unmappedVal in finalContext.argTypes[key].mapping
          ? finalContext.argTypes[key].mapping[unmappedVal]
          : unmappedVal
      
      acc[key] = Array.isArray(val) ? val.map(mappingFn) : mappingFn(val);

      return acc;
```